jobmaster/dm/metadata/state.go

```diff
@@ -0,0 +1,153 @@
+package metadata
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+	"sync"
+
+	"github.com/pingcap/errors"
+
+	"github.com/hanfei1991/microcosm/pkg/meta/metaclient"
+)
+
+// State represents the state which need to be stored in metadata.
+type State interface{}
+
+// Store manages a type of state.
+// Store provides the factory, some utility functions and persistence of State.
+type Store interface {
+	CreateState() State
+	Key() string
+}
+
+// DefaultStore implements some default methods of Store.
+type DefaultStore struct {
+	Store
+
+	state    State
+	kvClient metaclient.KVClient
+
+	mu sync.RWMutex
+}
+
+func NewDefaultStore(kvClient metaclient.KVClient) *DefaultStore {
+	return &DefaultStore{
+		kvClient: kvClient,
+	}
+}
+
+func (ds *DefaultStore) PutOp(state State) (metaclient.Op, error) {
+	v, err := json.Marshal(state)
+	return metaclient.OpPut(ds.Key(), string(v)), errors.Trace(err)
+}
+
+func (ds *DefaultStore) DeleteOp() metaclient.Op {
+	return metaclient.OpDelete(ds.Key())
+}
+
+// checkAllFieldsIsPublic check all fields of a pointer is public.
+func checkAllFieldsIsPublic(state State) bool {
+	v := reflect.ValueOf(state)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	if v.Kind() != reflect.Struct {
+		return false
+	}
+	for i := 0; i < v.NumField(); i++ {
+		if !v.Field(i).CanSet() {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
I am not familiar with reflect, these code are generated by Github Copilot, please review carefully